### PR TITLE
Add new admin tool, check bioeffects, complete with UI

### PR DESCRIPTION
**Pull request by Sovexe**
## About the PR 
Creates a new admin tool, Check Bioeffects complete with it's own interactive UI. Admins will be able to either use the tilde menu to select "Check Bioeffects" or the new Check link player options with a target selected. This will open a menu listing out all of their current bioeffects including both the bioeffect IDs and the friendly names. Additionally on the left of each effect a remove button is present, as you expect this removes the selected effect.

**EDIT:** Now includes the following additional features 
- search for and add bioeffects by path
- clickable interface features for controlling common verbs
- a splice menu for if you wanted to use the pre-made var combos that come with the chromosomes (including the option to make and name your own custom one!)

63fc0eade2c02aa7b8029f47254df936
b31a1d29dc897deadcb8871ed1bd6bf1
image
image
image
**OLD**
~~**video demo: https://streamable.com/alyewi**~~

~~https://user-images.githubusercontent.com/33204415/85974700-339e2a00-b9a4-11ea-88ec-dd224d72031a.png~~
~~https://user-images.githubusercontent.com/33204415/86092282-4a638000-ba7b-11ea-9bfd-37efa4cc712d.gif~~

## Why's this needed? 
There isn't really a good way to list out bioeffects. You can check the bioHolder.effects list but that only lists the IDs and isn't quickly accessible. This should serve as a nice QOL change to help make the lives of administration easier when working with bioeffects.

### CODE REVIEW

**Comment by Kyle2143** [comment]
This is pretty decent. I'd just inspect the bioholder list if I ever wanted to see em. As a suggestion, it would be good if you had an interface for altering the common chromosome vars for a bioeffect. I think they are: power, safety, synchronized.

**Comment by Gannets** [comment]
Very pretty!

**Comment by Sovexe** [comment]
> This is pretty decent. I'd just inspect the bioholder list if I ever wanted to see em. As a suggestion, it would be good if you had an interface for altering the common chromosome vars for a bioeffect. I think they are: power, safety, synchronized.

Sure thing

**Comment by Sovexe** [comment]
Now with chromosome modifying action!

video demo: https://streamable.com/k1sahi

**Comment by Sovexe** [comment]
-I added a stability readout
-You can now create custom chromosomes. You can pick a name (or not) and toggle the features of all ~main chromosomes (reinforcer, stab, power, synch) as well as specify a custom cooldown (entry defaults to the genes default cooldown)
-You can now search for and admin bioeffects right from the page

**Comment by Sovexe** [comment]
I thought it over and dislike how I did custom chromosomes. I’m going to change them to be in line clickable toggles

Converting this PR to a draft for now

**Comment by Sovexe** [comment]
https://gyazo.com/4d54738cba44fd82c59f0518d7582fde

**Comment by Sovexe** [comment]
@Kyle2143 I think you will find this latest update meets your original request much better than my custom chromosome.

4d54738cba44fd82c59f0518d7582fde

**Comment by Kyle2143** [comment]
Very nice, how much?

**Comment by Sovexe** [comment]
@Kyle2143 not just size but look in general. I think these buttons will be much more user friendly, I also remove the underlines from the links in the rows so they look cleaner. Here's a clip of the most up to date UI

image

**Comment by Tarmunora** [comment]
Perhaps include the option for timed effects and making a bioeffect "magic" (prevents genetics console interaction iirc)?

**Comment by Sovexe** [comment]
> Perhaps include the option for timed effects and making a bioeffect "magic" (prevents genetics console interaction iirc)?

making effects magical forces several vars and is still accessible if adding genes through the main Add button in player options. I purposely chose not to have this apply as magic as I wanted genes to be added in in their default state.

I do not believe magic prevents storing of genes, looking at the code magical bioeffects have the following vars preset, the genes would be non-reclaimable, and unable to be lost to a DNA scramble

```				newEffect.curable_by_mutadone = 0
				newEffect.stability_loss = 0
				newEffect.can_scramble = 0
				newEffect.can_reclaim = 0

**Comment by Sovexe** [comment]
the build errors appear to be an issue with the checks themselves
```Reading package lists...
E: Failed to fetch https://storage.googleapis.com/bazel-apt/dists/stable/jdk1.8/binary-amd64/Packages.gz  File has unexpected size (3690 != 3912). Mirror sync in progress? [IP: 172.217.13.80 443]
   Hashes of expected file:
    - Filesize:3912 [weak]
    - SHA256:0cf48c3fab1fc1b2c564d966d2c3bf307e73c734a841315e545e261f7c858b72
    - SHA1:c4745b298dc8092ee275faa5580331e11f808222 [weak]
    - MD5Sum:a83724d031b3d66d0000caa3aa4d55f8 [weak]
   Release file created at: Tue, 30 Jun 2020 15:25:43 +0000
E: Some index files failed to download. They have been ignored, or old ones used instead.
##[error]Process completed with exit code 100.```

**Comment by mordent-goonstation** [comment]
Disagree with underline removal on links - the underlines are an important UX hint to users. An alternative is to make them checkboxes etc., but that doesn't work as nicely out-of-the-box so would require more work (so I'd just keep it as links for now).

**Comment by mordent-goonstation** [comment]
Unclear if the feature is on here based on a cursory look (not examined the code in detail), but could you make the stability of the user editable if it isn't?

**Comment by Sovexe** [comment]
@mordent-goonstation I have re-added the underscores and added the ability to edit stability

image

File: code/modules/admin/admin.dm [hunk 1]
```diff
@@ -4224,6 +4430,125 @@ var/global/noir = 0
...
+	if (!BE.stability_loss) //reapply stability changes
+		BE.stability_loss = BE.global_instance.stability_loss
+		BE.holder.genetic_stability = max(0, BE.holder.genetic_stability -= BE.stability_loss)
+	BE.curable_by_mutadone = BE.global_instance.curable_by_mutadone
+	BE.reclaim_fail = BE.global_instance.reclaim_fail
+	BE.reclaim_mats = BE.global_instance.reclaim_mats
+	BE.msgGain = BE.global_instance.msgGain
+	BE.msgLose = BE.global_instance.msgLose
+	if (istype(BE, /datum/bioEffect/power)) //powers
+		BE:power = BE.global_instance:power
```
Comment (mordent-goonstation) [inline code/modules/admin/admin.dm:4446]: Prefer casting to using `:`.

**Comment by Sovexe** [comment]
@mordent-goonstation all `:` should now be replaced with `.` and type casting

it was a little tricky to do `proc/checkbioeffect_chromosome_clean` as the var `global_instance` I was using to reset them is specifically a `/datum/bioEffect`. As such I ended making a new var on `datum/BioEffect` named `global_instance_power` that gets populated in the `New` proc if the bioeffect passes an isType check.

**Comment by ZeWaka** [comment]
this is definitely a very good canidate for early tgui adoption, checkbox/button elements would be very useful here

**Pull request merged**